### PR TITLE
repo: add multiarch support for stage packages.

### PR DESCRIPTION
**Pull request by kyrofa**
Resolve LP: [#1655797](https://bugs.launchpad.net/snapcraft/+bug/1655797) by adding support for staging packages meant for another architecture with the 'name:arch' syntax. Note that the build machine needs to be configured to pull packages for that architecture, though (`dpkg --add-architecture`).

### CODE REVIEW

**Comment by sergiusens** [comment]
hmmm, I guess you missed the comment on the mailing list where I said I would fix this. It would be ideal to focus on the other cases. There is a simpler solution for this.

**Comment by kyrofa** [comment]
> hmmm, I guess you missed the comment on the mailing list where I said I would fix this. It would be ideal to focus on the other cases. There is a simpler solution for this.

Indeed I did. Alright, I'll assign it to you.

File: snapcraft/tests/test_repo.py [hunk 1]
```diff
@@ -70,6 +71,37 @@ def test_get_package(self, mock_apt):
...
+            call.Cache().open(),
+        ])
+        mock_apt.assert_has_calls([
+            call.Cache().fetch_archives(progress=ANY),
+        ])
+
+        # __getitem__ is tricky
+        self.assertThat(
+            mock_apt.Cache().__getitem__.call_args_list,
+            Contains(call('fake-package:arch')))
```
Comment (sergiusens) [inline snapcraft/tests/test_repo.py]: ok, this seems to be the way :-)

```
In [7]: p = a['bash']

In [9]: p.architecture()
Out[9]: 'amd64'

In [10]: p = a['bash:i386']

In [12]: p.architecture()
Out[12]: 'i386'
```

File: integration_tests/test_multiarch_stage_packages.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    def test_bad_multiarch_stage_packages(self):
+        project_dir = 'bad-arch'
+        exception = self.assertRaises(
+            subprocess.CalledProcessError, self.run_snapcraft,
+            'pull', project_dir)
+
+        self.assertThat(
+            exception.output, Contains(
+                "Error downloading stage packages for part 'my-part': The "
+                "Ubuntu package 'hello:fake-arch' was not found. You may need "
```
Comment (sergiusens) [inline integration_tests/test_multiarch_stage_packages.py]: Can you add a newline after the `.` and before the `You`? Would make it much more readable on a console

Comment (kyrofa) [inline integration_tests/test_multiarch_stage_packages.py]: Yeah good call-- done.

**Comment by mardy** [comment]
Filed a bug about the same feature, just for build-packages:
https://bugs.launchpad.net/snapcraft/+bug/1664141

**Pull request merged**